sycl/include/CL/sycl/intel/group_algorithm.hpp

```diff
@@ -0,0 +1,660 @@
+  id<3> result;
+  result[0] = i / (r[1] * r[2]);
+  result[1] = (i % (r[1] * r[2])) / r[2];
+  result[2] = i % r[2];
+  return result;
+}
+
+template <typename T> struct is_group : std::false_type {};
+
+template <int Dimensions>
+struct is_group<group<Dimensions>> : std::true_type {};
+
+template <typename T, class BinaryOperation> struct identity {};
+
+template <typename T, typename V> struct identity<T, intel::plus<V>> {
+  static constexpr T value = 0;
+};
+
+template <typename T, typename V> struct identity<T, intel::minimum<V>> {
+  static constexpr T value = std::numeric_limits<T>::max();
+};
+
+template <typename T, typename V> struct identity<T, intel::maximum<V>> {
+  static constexpr T value = std::numeric_limits<T>::lowest();
+};
+
+template <typename Group, typename Ptr, class Function>
+Function for_each(Group g, Ptr first, Ptr last, Function f) {
+#ifdef __SYCL_DEVICE_ONLY__
+  nd_item<Group::dimensions> it =
+      ::sycl::detail::Builder::getNDItem<Group::dimensions>();
+  ptrdiff_t offset = it.get_local_linear_id();
+  ptrdiff_t stride = detail::get_local_linear_range(g);
+  for (Ptr p = first + offset; p < last; p += stride) {
+    f(*p);
+  }
+  return f;
+#else
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_INVALID_DEVICE);
+#endif
+}
+
+} // namespace detail
+
+namespace intel {
+
+template <typename T>
+using EnableIfIsScalarArithmetic =
+    ::sycl::detail::enable_if_t<::sycl::detail::is_scalar_arithmetic<T>::value,
+                                T>;
+
+template <typename T>
+using EnableIfIsVectorArithmetic =
+    ::sycl::detail::enable_if_t<::sycl::detail::is_vector_arithmetic<T>::value,
+                                T>;
+
+template <typename Ptr, typename T>
+using EnableIfIsPointer =
+    ::sycl::detail::enable_if_t<::sycl::detail::is_pointer<Ptr>::value, T>;
+
+template <typename Group> bool all_of(Group g, bool pred) {
+  static_assert(detail::is_group<Group>::value,
+                "Group algorithms only support the sycl::group class.");
+#ifdef __SYCL_DEVICE_ONLY__
+  return __spirv_GroupAll(__spv::Scope::Workgroup, pred);
+#else
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_INVALID_DEVICE);
+#endif
+}
+
+template <typename Group, typename T, class Predicate>
+bool all_of(Group g, T x, Predicate pred) {
+  static_assert(detail::is_group<Group>::value,
+                "Group algorithms only support the sycl::group class.");
+  return all_of(g, pred(x));
+}
+
+template <typename Group, typename Ptr, class Predicate>
+EnableIfIsPointer<Ptr, bool> all_of(Group g, Ptr first, Ptr last,
+                                    Predicate pred) {
+#ifdef __SYCL_DEVICE_ONLY__
+  static_assert(detail::is_group<Group>::value,
+                "Group algorithms only support the sycl::group class.");
+  bool partial = true;
+  detail::for_each(g, first, last, [&](auto x) { partial &= pred(x); });
```

Comment:
` error: 'auto' not allowed in lambda parameter`

Looks like it should be protected with C++ version check.

Reply:
This is a C++14 feature, but since that's the default selected by Clang I thought it would be okay to use. I see /usr/bin/c++ being used in the Buildbot output -- what compiler, and what minimum version of C++, should I be targeting?

Also, this same code worked a few days ago.  Did something change?

Reply:
I think adding `group_algorithm.hpp` header to `sycl.hpp` broke the host-only compilation.
NOTE: Current version of CTS validates conformance to SYCL-1.2.1, which requires C++11 support.

Reply:
Ah, I didn't realize that.  I'll start testing with C++11 locally going forward.

Reply:
Wait what?
CTS requiring C++11 should only mean that nothing in the CTS code and APIs used by CTS use more than C++11.  That should not preclude -std=c++14 at all.

Reply:
> CTS requiring C++11 should only mean that nothing in the CTS code and APIs used by CTS use more than C++11.

This patch violates this requirement by adding C++14 code via SYCL headers to CTS code.